### PR TITLE
addresses a possible race condition with purchase completed callbacks

### PR DESCRIPTION
**Pull request by aboedo**
There was a race condition in `handleReceiptPostWithTransaction:purchaserInfo:subscriberAttributes:error:`. 

The issue was that in the order of operations, we were: 

1. fetching the completion block
1. performing the completion block
1. removing it from the callbacks list. 

If the completion block performed in 2. does something like initiate another purchase, it would access the callbacks list  before it got removed in 3, and then the code would see that there's already a completion block for the product and return the `RCOperationAlreadyInProgressError`. 

might address #291, but it's not guaranteed to since I'm not sure that this was the original issue.

### CODE REVIEW

File: PurchasesTests/Purchasing/PurchasesTests.swift [hunk 1]
```diff
@@ -1972,6 +1972,37 @@ class PurchasesTests: XCTestCase {
...
+    func testPurchasingPackageDoesntThrowPurchaseAlreadyInProgressIfCallbackMakesANewPurchase() {
+        setupPurchases()
+        var receivedError: NSError? = nil
+        var secondCompletionCalled = false
+        self.purchases!.offerings { (newOfferings, _) in
+            let package = newOfferings!["base"]!.monthly!
+            self.purchases!.purchasePackage(package) { _,_,_,_  in
+                self.purchases!.purchasePackage(package) { (tx, info, error, userCancelled) in
+                    receivedError = error as NSError?
+                    secondCompletionCalled = true
```
Comment (aboedo) [inline PurchasesTests/Purchasing/PurchasesTests.swift:1985]: this is the error case - while the completion block is running, it hasn't been removed from the list of completion blocks, and so initiating a new purchase will fail. 
I confirmed that this test fails on `develop`

Comment (vegaro) [reply PurchasesTests/Purchasing/PurchasesTests.swift:1985]: good catch!

File: Purchases/Public/RCPurchases.m [hunk 2]
```diff
@@ -1041,10 +1028,11 @@ - (void)storeKitWrapper:(RCStoreKitWrapper *)storeKitWrapper
 
 - (nullable RCPurchaseCompletedBlock)getAndRemovePurchaseCompletedBlockFor:(SKPaymentTransaction *)transaction {
     RCPurchaseCompletedBlock completion = nil;
-    if ([self productIdentifierFrom:transaction]) {
+    NSString * _Nullable productIdentifier = [self productIdentifierFrom:transaction];
+    if (productIdentifier) {
         @synchronized (self) {
-            completion = self.purchaseCompleteCallbacks[transaction.payment.productIdentifier];
-            self.purchaseCompleteCallbacks[transaction.payment.productIdentifier] = nil;
```
Comment (vegaro) [inline Purchases/Public/RCPurchases.m:1047]: thanks for fixing the productIdentifier thing, that was bad from me

File: Purchases/Public/RCPurchases.m [hunk 3]
```diff
@@ -968,12 +961,6 @@ - (void)handleReceiptPostWithTransaction:(SKPaymentTransaction *)transaction
             RCLog(@"Unexpected error from backend");
             CALL_IF_SET_ON_SAME_THREAD(completion, transaction, nil, error, false);
         }
-
-        @synchronized (self) {
-            if (productIdentifier) {
-                self.purchaseCompleteCallbacks[productIdentifier] = nil;
-            }
-        }
```
Comment (vegaro) [inline Purchases/Public/RCPurchases.m:976]: It's funny, this is exactly what you said the other day that could break

**Pull request merged**